src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java

```diff
@@ -264,38 +262,48 @@ private String bitbucketUrl() {
         return StringUtils.defaultIfBlank(bitbucketServerUrl, "https://bitbucket.org");
     }
 
-    public void setBitbucketConnector(@NonNull BitbucketApiConnector bitbucketConnector) {
-        this.bitbucketConnector = bitbucketConnector;
-    }
-
-    BitbucketApiConnector getBitbucketConnector() {
-        if (bitbucketConnector == null) {
-            bitbucketConnector = new BitbucketApiConnector(bitbucketServerUrl);
+    public String getRemote(@NonNull String repoOwner, @NonNull String repository, BitbucketRepositoryType repositoryType) {
+        assert repositoryType != null;
+        BitbucketRepositoryProtocol protocol;
+        Integer protocolPortOverride = null;
+        if (StringUtils.isBlank(checkoutCredentialsId)) {
+            protocol = BitbucketRepositoryProtocol.HTTP;
+        } else if (getCheckoutCredentials() instanceof SSHUserPrivateKey) {
+            protocol = BitbucketRepositoryProtocol.SSH;
+            if (sshPort > 0) {
+                protocolPortOverride = sshPort;
+            }
+        } else {
+            protocol = BitbucketRepositoryProtocol.HTTP;
         }
-        return bitbucketConnector;
+        return buildBitbucketClient().getRepositoryUri(repositoryType, protocol, protocolPortOverride, repoOwner, repository);
     }
 
-    public String getRemote(@NonNull String repoOwner, @NonNull String repository) {
-        return getUriResolver().getRepositoryUri(repoOwner, repository, getRepositoryType());
-    }
-
-    public RepositoryType getRepositoryType() {
+    public BitbucketRepositoryType getRepositoryType() throws IOException, InterruptedException {
         if (repositoryType == null) {
-            BitbucketRepository r = getBitbucketConnector().create(repoOwner, repository, getScanCredentials()).getRepository();
-            if (r == null) {
-                throw new AssertionError("Not found repository: " + repoOwner + "/" + repository);
-            }
-            repositoryType = RepositoryType.fromString(r.getScm());
+            repositoryType = BitbucketRepositoryType.fromString(buildBitbucketClient().getRepository().getScm());
         }
         return repositoryType;
     }
 
     public BitbucketApi buildBitbucketClient() {
-        return getBitbucketConnector().create(repoOwner, repository, getScanCredentials());
+        return BitbucketApiFactory.newInstance(bitbucketServerUrl, getScanCredentials(), repoOwner, repository);
     }
 
     public BitbucketApi buildBitbucketClient(PullRequestSCMHead head) {
-        return getBitbucketConnector().create(head.getRepoOwner(), head.getRepository(), getScanCredentials());
+        return BitbucketApiFactory.newInstance(bitbucketServerUrl, getScanCredentials(), head.getRepoOwner(), head.getRepository());
+    }
+
+    @Override
+    public void afterSave() {
```

Comment:
Is this defensive? I think it shouldn't happen (unless there is some connectivity issue)